lib/cli.js

```diff
@@ -207,4 +207,7 @@ exports.run = () => {
 				});
 			});
 	}
+
+	// Trap SIGINT to prevent AVA quitting immediately
+	process.on('SIGINT', () => {});
```

Comment:
This relies on the signal still being propagated to the worker processes, right? I think we need to take a more explicit approach. For one the API needs to know not to start new workers when concurrency is being limited. Since the API is managing the worker pool it could shut down the workers, and the workers could then emit their pending tests. See https://github.com/avajs/ava/blob/e401bd151b1a81f79ccb106f034b2e18715ee125/api.js#L245:L247

(Though looking at that code I have a suspicion timeouts don't prevent new workers from being started!)

We'd need to make sure the signal isn't propagated, or otherwise is ignored in the workers.

Reply:
The teardown process is initiated by the `SIGINT` handler added to `lib/main.js` - which then calls the `exit()` function in that same file.

If you disable the `SIGINT` handler in `lib/main.js` while keeping this one, AVA (from what I can see through the CLI) basically ignores `SIGINT`s. If you disable this `SIGINT` handler however, `lib/main.js` doesn't have enough time to gracefully exit.

This was largely based on experimenting with the code though - so it's definitely possible that I'm initiating the wrong teardown call/event.

Reply:
I'd like to make sure we try intercepting `SIGINT` in the main process, and then using the `Api` instance to explicitly exit all workers. We may then need to trap `SIGINT` in the workers to ensure they don't exit before they're told to do so.